src/gui/screen_bed_leveling.cpp

```diff
@@ -0,0 +1,85 @@
+#include "ScreenHandler.hpp"
+#include "gui.hpp"
+
+screen_bed_leveling_t::screen_bed_leveling_t()
+    : AddSuperWindow<screen_t>()
+    , m_statusMsg(this, { 10, 128, 220, 80 }, is_multiline::no)
+    , m_progressBar(this, Rect16(10, 70, GuiDefaults::RectScreen.Width() - 2 * 10, 16))
+    , header(this)
+    , footer(this) {
+    header.SetIcon(IDR_PNG_heatbed_16px);
+    header.SetText(_("BED LEVELING"));
+    this->m_statusMsg.SetText(_(""));
+}
+
+void screen_bed_leveling_t::windowEvent(EventLock /*has private ctor*/, window_t *sender, GUI_event_t event, void *param) {
+    if (event == GUI_event_t::LOOP) {
+        switch (this->state) {
+        case state_t::idle: // Start homing
+            // Start by clearing events and homing
+            marlin_error_clr(MARLIN_ERR_ProbingFailed);
+            marlin_event_clr(MARLIN_EVT_CommandBegin);
+            marlin_event_clr(MARLIN_EVT_CommandEnd);
+            marlin_gcode("G28");
+            this->m_statusMsg.SetText(_("Homing..."));
+            while (!marlin_event_clr(MARLIN_EVT_CommandBegin))
+                marlin_client_loop();
+            this->state = state_t::homing;
+            break;
+
+        case state_t::homing:
+            // If homing is finished, start mesh bed leveling
+            if (marlin_event_clr(MARLIN_EVT_CommandEnd)) {
+                marlin_gcode("G29");
+                while (!marlin_event_clr(MARLIN_EVT_CommandBegin))
+                    marlin_client_loop();
+                this->state = state_t::leveling;
+                this->updateStatusMsg();
+            }
+            break;
+
+        case state_t::leveling: {
+            this->updateStatusMsg();
+
+            // If leveling is finished, jump the an end state
+            if (marlin_event_clr(MARLIN_EVT_CommandEnd)) {
+                if (marlin_error(MARLIN_ERR_ProbingFailed) || m_currentLevelingPointNum != m_levelingPointCount) {
+                    this->state = state_t::error;
+                } else {
+                    this->state = state_t::done;
+                }
+            }
+            break;
+        }
+        case state_t::done:
+            Screens::Access()->Close();
+            break;
+        case state_t::error:
+            marlin_error_clr(MARLIN_ERR_ProbingFailed);
+            m_statusMsg.SetText(_("Bed leveling failed"));
+            if (!m_errorSleepCycles--)
+                Screens::Access()->Close();
+            break;
+        default:
+            Screens::Access()->Close();
+            break;
+        }
+    }
+
+    SuperWindowEvent(sender, event, param);
+}
+
+void screen_bed_leveling_t::updateStatusMsg() {
+    // Update status message
+    const char *msg = MsgCircleBuffer().GetLastIfAble();
+    if (msg) {
+        const uint8_t *data = reinterpret_cast<const uint8_t *>(msg);
+        auto newText = string_view_utf8::MakeRAM(data);
+        if (newText != this->m_statusMsg.GetText()) {
+            this->m_statusMsg.SetText(newText);
```

Comment:
The data could change under our hands between the call and invalidate. 
MakeRAM does not allocate any memory, it uses the memory where the pointer points.

Reply:
This should be fixed by keeping an internal copy.